pointpats/spacetime.py

```diff
@@ -629,3 +658,1008 @@ def _shuffle_matrix(X, ids):
+                    "number. Using analytic p-values instead"
+                )
+                g["pvalue"] = self.p_hypergeom
+            else:
+                g["pvalue"] = self.p_sims
+        elif inference == "analytic":
+            g["pvalue"] = self.p_hypergeom
+        else:
+            raise ValueError("inference must be either `permutation` or `analytic`")
+
+        mask = g[g.pvalue <= crit].index.values
+        neighbors = self.adjlist[self.adjlist.focal.isin(mask)].neighbor.unique()
+
+        # this is clunky, but enforces plotting order so significance is prioritized
+        g.loc[neighbors, "color"] = colors["neighbor"]
+        g.loc[g.pvalue <= crit, "color"] = colors["focal"]
+        nbs = self.adjlist.groupby("focal").agg(list)["neighbor"]
+        g = g.reset_index().merge(nbs, left_on="index", right_index=True, how="left")
+
+        m = g[g.color == colors["nonsig"]].explore(
+            color="grey", style_kwds=style_kwds, tiles=tiles
+        )
+        blues = g[g.color == colors["neighbor"]]
+        if blues.shape[0] == 0:
+            warn("empty neighbor set.")
+        else:
+            m = blues.explore(m=m, color=colors["neighbor"], style_kwds=style_kwds)
+        m = g[g.color == colors["focal"]].explore(
+            m=m, color=colors["focal"], style_kwds=style_kwds
+        )
+
+        if plot_edges:
+
+            # edges between hotspot and st-neighbors
+            g = g.set_index("index")
+            ghs = self.hotspots(crit=crit, inference=inference)
+            ghs = ghs.dropna()
+            origins = g.loc[ghs.focal].geometry
+            destinations = g.loc[ghs.neighbor].geometry
+            ods = zip(origins, destinations)
+            lines = gpd.GeoSeries([LineString(od) for od in ods])
+            lines.crs = g.crs
+            lines.explore(m=m, color=edge_color, style_kwds={"weight": edge_weight})
+
+        return m
+
+    def _gdfhs(self, crit=0.05, inference="permutation"):
+        # merge df with self.hotspots
+        return gpd.GeoDataFrame(
+            self._gdf.merge(
+                self.hotspots(crit=crit, inference=inference),
+                left_index=True,
+                right_on="focal",
+            )
+        )
+
+
+def _spacetime_points_to_arrays(dataframe, time_col):
+    """convert long-form geodataframe into arrays for kdtree
+
+    Parameters
+    ----------
+    dataframe : geopandas.GeoDataFrame
+        geodataframe with point geometries
+    time_col : str
+        name of the column on dataframe that stores time values
+
+    Returns
+    -------
+    tuple
+        two numpy arrays holding spatial coodinates s_coords (n,2)
+        and temporal coordinates t_coords (n,1)
+
+    """
+    if dataframe.crs is None:
+        warn(
+            "There is no CRS set on the dataframe. The KDTree will assume coordinates "
+            "are stored in Euclidean distances"
+        )
```

Comment:
just fyi, GeoPandas does not warn in this case. We only warn when the crs is geographic. No CRS implies cartesian coordinates.

Reply:
we hit this with the burkitt data which is stored in planar coords but doesnt have a CRS, so figured i'd add a warning just in case. I dont think we can always assume no CRS necessarily implies cartesian coords here.

Reply:
I don't mind the warning, just wanted to clarify the assumptions geopandas makes in case we want to do the same.

Reply:
ah gotcha. I opted to warn here because the KDtree will still "work" (and thus the class will produce a result), but afaik there's no way to make it work *correctly* with cartesian coords, so its up to the user if they know what they're doing. Maybe it would be more conservative to fail instead?